FileCopy/files.cpp

```diff
@@ -0,0 +1,214 @@
+    assert(fs && prep_out && sections_out && check_out);
+    const char *filename = fs->files[id].filename;
+    assert(filename);
+
+    uint8_t *buffer = nullptr;
+    checksum_t checksum;
+
+    // read the file
+    char fullname[FULLNAME];
+    mkfullname(fullname, fs->dirname, filename);
+    assert(is_file(fullname) && fs->nfp->fopen(fullname, "rb"));
+
+    int len = fmemread_secure(fs->nfp, fs->nastiness, &buffer, checksum);
+    assert(len > 0);
+    fs->nfp->fclose();
+
+    const int max_data = sizeof((*sections_out)->value.section.data);
+    int nparts = (len / max_data) + (len % max_data ? 1 : 0);
+
+    // make prepare_for_blob message packet
+    packet_prepare(prep_out, id, filename, len, nparts);
+
+    // make all blob_section packets
+    packet_t *sections = (packet_t *)malloc(sizeof(packet_t) * nparts);
+    assert(sections);
+    for (int i = 0; i < nparts; i++) {
+        int offset = i * max_data;
+        int size = min(max_data, len - offset);
+        packet_section(&sections[i], id, i, offset, size, buffer + offset);
+    }
+
+    // make check is necessary packet
+    packet_checkisnecessary(check_out, id, filename, checksum);
+
+    free(buffer);
+    *sections_out = sections;
+    return nparts;
+}
+
+// buffer_in != nullptr && 0 <= start <= end <= length
+// guaranteed safe!
+//
+// stores in TMP file
+bool files_writetmp(files_t *fs, int id, int nbytes, const void *buffer_in,
+                    const checksum_t checksum_in) {
+    assert(buffer_in && fs);
+    assert(nbytes > 0);
+
+    // file open
+    char tmpname[FULLNAME];
+    mkfullname(tmpname, fs->dirname, mktmpname(fs->files[id].filename));
+    assert(fs->nfp->fopen(tmpname, "w+b"));
+
+    checksum_t checksum;
+
+    int attempts = 0;
+    do {
+        uint8_t *buffer = nullptr;
+        fs->nfp->fseek(0, SEEK_SET);
+        fs->nfp->fwrite(buffer_in, 1, nbytes);
+        int len = fmemread_secure(fs->nfp, fs->nastiness, &buffer, checksum);
+        assert(len == nbytes);
+        free(buffer);
+        if (memcmp(checksum_in, checksum, SHA_LEN) == 0) {
+            fs->nfp->fclose();
+            return true;
+        }
+    } while (++attempts < DISK_RETRIES(fs->nastiness));
+
+    errp("Total Disk Failure sending SOS\n");
+    fs->nfp->fclose();
+    return false;
+}
+
+// renames TMP file to permanent
+void files_savepermanent(files_t *fs, int id) {
+    char filename[FULLNAME];
+    char tmpname[FULLNAME];
+    mkfullname(filename, fs->dirname, fs->files[id].filename);
+    mkfullname(tmpname, fs->dirname, mktmpname(fs->files[id].filename));
+    rename(tmpname, filename);
+}
+
+void files_remove(files_t *fs, int id) { remove(fs->files[id].filename); }
+
+// Brute force secure read
+// Needs to have <nastiness> matching reads
+int fmemread_secure(C150NastyFile *nfp, const int nastiness,
+                    uint8_t **buffer_pp, unsigned char checksum_out[]) {
+    verify(buffer_pp && *buffer_pp == nullptr);
```

Comment:
Should this be an assertion? If `buffer_pp` isn't in the right state isn't that a bug?